File: flamingo-tests/mobile-reference/FrameSource.kt
package com.flamingo.sedum.source

import kotlinx.coroutines.flow.Flow

/** Кадр: буфер пикселей + монотонная метка времени. */
data class Frame(val pixels: ByteArray, val width: Int, val height: Int, val timestampNanos: Long)

/**
 * SEAM. Единственное место в проекте, знающее про androidx.camera.
 * Выше по стеку (VisionPipeline, UI) камера не импортируется никогда.
 */
interface FrameSource {
    fun frames(): Flow<Frame>
    fun stop()
}

// --- ПРОД ---
// class LiveCameraSource(...) : FrameSource   // androidx.camera / CameraX

// --- ТЕСТЫ: чтение кадров из фикстурного видео по виртуальным часам ---
// class FileFrameSource(assetName: String, private val clock: VirtualClock) : FrameSource

/**
 * Выбор источника по launch-аргументу:
 *   adb shell am instrument -e frame_source fixture_looking_away ...
 *
 * object FrameSourceFactory {
 *     fun create(args: Bundle?): FrameSource {
 *         val fixture = args?.getString("frame_source")
 *         return if (fixture != null) FileFrameSource("fixtures/$fixture.mp4", VirtualClock())
 *                else LiveCameraSource()
 *     }
 * }
 */

File: flamingo-tests/mobile-reference/frameSource.ts
/** Вариант для React Native. */

export interface Frame {
  data: Uint8Array;
  width: number;
  height: number;
  timestamp: number;   // монотонные мс
}

/** SEAM: единственный модуль, знающий про react-native-vision-camera. */
export interface FrameSource {
  frames(): AsyncIterable<Frame>;
  stop(): void;
}

// --- ПРОД ---
// export class LiveCameraSource implements FrameSource { /* vision-camera frameProcessor */ }

// --- ТЕСТЫ: декодирование фикстуры по виртуальным часам ---
// export class FileFrameSource implements FrameSource {
//   constructor(private fixture: string, private clock: VirtualClock) {}
// }

export function createFrameSource(): FrameSource {
  // Значение прокидывается launch-аргументом / переменной окружения сборки
  const fixture = process.env.FLAMINGO_FRAME_SOURCE;
  if (fixture) {
    throw new Error(`подставьте FileFrameSource('${fixture}')`);
  }
  throw new Error('подставьте LiveCameraSource()');
}
